tests/ci/check_so_interface.py

```diff
@@ -276,68 +303,94 @@
+        raise Exception("Kwarg expected to be found in "+ (node.name) + " but kwarg not found")
+    #If kwargs not detected but kwargs expected
+    if not has_kwargs and not intdict.get('kwarg') == None:
+        raise Exception("Kwarg not expected to be found in " + (node.name) + " but kwarg detected")
+    # Compare arguments to specified arguments in interface dict
+    if found_args != spec_args:
+        print(f"Mismatch in arguments for function {node.name}")
+        print(f"Found: {found_args}")
+        print(f"Expected: {spec_args}")
+        return "incorrectargs"
+    # Compare default values to specified defaults in interface dict
+    if found_defaults != spec_defaults:
+        print(f"Mismatch in default values for function {node.name}")
+        print(f"Found: {found_defaults}")
+        print(f"Expected: {spec_defaults}")
+        return "incorrectdefaults"
     return True
 
 if __name__ == "__main__":   #if this is the main thing being run
-    setspecclass = set()
-    setspecfunc = set()
+    setspecclass = set() #iterate over dict
+    setspecfunc = set() #iterate over dict
     foundclass = set()
     foundfunc = set()
     for fname in interface.keys():   #loop over file names "fname" as keys in dictionary called "interface"
+        for specclass in interface[fname].keys():
+            setspecclass.add(specclass)
+            for specfunc in interface[fname][specclass].keys():
+                setspecfunc.add(specfunc)
         with open(fname, 'r') as fh:   #opens file "fname", in read mode 'r', to be used in code as "fh"
             tree = ast.parse(fh.read()) #parsing contents of file into abstract syntax tree
-            dump = ast.dump(tree)
-        for classname in interface[fname].keys():
-            setspecclass.add(classname)
-            for node in ast.walk(tree):
-                if isinstance(node, ast.ClassDef):
-                    if node.name == classname:
-                        intdict = interface[fname][classname]
-                        assert compare_args(node, intdict, classname)
-                        foundclass.add(classname)
-                for specfunc in interface[fname][classname].keys():
-                    setspecfunc.add(specfunc)
-                    if isinstance(node, ast.FunctionDef) and node.name == specfunc:
-                        intdict = interface[fname][classname][specfunc]
-                        assert compare_args(node, intdict, specfunc)
-                        foundfunc.add(specfunc)
+        for node in ast.walk(tree):
+            if isinstance(node, ast.ClassDef):
+                # check if this class is in our spec
+                if node.name in interface[fname]:
+                    # check the class
+                    foundclass.add(node.name)
+                    for child in node.body:
+                        if isinstance(child, ast.FunctionDef) and child.name in interface[fname][node.name]:
```

Comment:
If the only behaviour you expect is an exception to be raised on failure, this can also be simplified a bit using the `try/except`syntax.
```suggestion
                        if isinstance(child, ast.FunctionDef) and child.name in interface[fname][node.name]:
                            try:
                                # this will function will raise on exception on failure
                                assert compare_args(child, interface[fname][node.name][child.name])
                            except Exception as e:
                                # if an exception was raised this block is executed
                                # we raise a new exception with some more context, based on the original one
                                raise Exception("Function check failed in file: " + fname) from e
```
Note that the `raise .. from` syntax will result in the caught exception being printed as well and the new one containing the information about the file name.